### DISCUSSION TITLE
Getting (reading) metadata from live stream

### DISCUSSION THREAD

**Discussion by gavin2812** [savonet/liquidsoap#3230] [User Support]
Hello,

I have seen quite a few examples in the documentation about setting/updating the metadata using insert_metadata but I can't immediately find anything for reading the metadata from a live stream and was wondering if anyone could help...

This is what I'm thinking **as an example**, I know code this does not work, but what I'm trying to do is to get the metadata (artist - title) every 20 seconds and output it via print (every 20 seconds):

```
url = "http://livestream.com/stream"
input = mksafe(input.http(url))

input = metadata(window=20., input)

thread.run(
  every=20.,
  {
  value := input.metadata()
  }
)

thread.run(
  every=20.,
  {
  print (value)
  }
)
```

Could someone help with this? I'm just trying to print out the artist - title of the live stream.

Thanks in advance.

**Comment by smimram** [comment]
You should use `on_metadata` on the source and store it in a reference when there is new metadata.

**Reply by gavin2812** [reply]
@smimram thanks for your quick reply, appreciated.

I found a couple of demo scripts and managed to get it working, thanks!

**Reply by gavin2812** [reply]
@smimram I'm now able to print the artist/title to console, but I was hoping to use this curl line (which I have working elsewhere in another script and it works) but by the time it gets to the destination server, the {title} and {artist} fields are empty.

def print_meta(m) =
	title = m["title"]
  	artist = m["artist"]
  	print("#{artist} - #{title}")
  	ignore(process.run("curl https://server.com/api?artist=#{artist}'&'title=#{title}"))
end

The `print("#{artist} - #{title}")` does correctly output the content to console but by the time it hits the server the artist and title fields are empty:

`xxx.xxx.xxx.xxx - - [14/Jul/2023:11:48:26 +0100] "GET /api/index.php?artist=&title= HTTP/1.1" 200 5684 "-" "curl/7.74.0"`

I've tried a few different things, but it either throws an error or just doesn't work.

Do you know how I can do this?

Thanks

**Comment by gavin2812** [comment]
So I now have it working, but there is a problem whereby I need to url.encode the data, I though I could use url.encode but it is giving me an error:

Updated code:
```
def print_meta(m) =
	title = url.encode(m["title"])
  	artist = url.encode(m["artist"])
  	  	  	
  	print("#{artist} - #{title}")  	
  	ignore(process.run("curl https://server.com/api?artist=#{artist}'&'title=#{title}
end
```

Gives this error:

```
/usr/bin/liquidsoap /home/test.liq
INFO: Loading Sdl_ttf, Target = linux
INFO: Loading Sdl_image, Target = linux
At /home/test/toolbox.liq, line 9, char 9-12:
	title = url.encode(m["title"])

Error 5: this value has no method `encode`
```
I'm assuming I am not using url.encode correctly?

Thanks.

**Reply by smimram** [reply]
This means that `url.encode` is not defined. This can be because
- you are using an "old" Liquidsoap version in which the function was not introduced yet
- you are defining a variable `url` above in the script, which hides the `url` object from the standard library

**Comment by Moonbase59** [comment]
@gavin2812: Just gave it a quick spin here, using `python3 -m http.server` and Liquidsoap 2.2.0+git@0a90fcf99 for a quick test:

```ruby
a = mksafe(playlist("~/Musik/Electronic.m3u"))

def print_meta(m)
  artist = url.encode(m["artist"])
  title = url.encode(m["title"])
  print("#{artist} - #{title}")
  ignore(process.run("curl http://localhost:8000/api?artist=#{artist}'&'title=#{title}"))
end

a.on_metadata(print_meta)

output(a)
```

Works fine:
```
127.0.0.1 - - [14/Jul/2023 17:36:53] "GET /api?artist=Klaus+Schulze&title=Blue+Spirits HTTP/1.1" 404 -
127.0.0.1 - - [14/Jul/2023 17:40:55] code 404, message File not found
127.0.0.1 - - [14/Jul/2023 17:40:55] "GET /api?artist=IC&title=Mann+im+Mond HTTP/1.1" 404 -
127.0.0.1 - - [14/Jul/2023 17:44:04] code 404, message File not found
127.0.0.1 - - [14/Jul/2023 17:44:04] "GET /api?artist=Sneaker+Pimps&title=Kiro+TV HTTP/1.1" 404 -
```

(404 error since I actually got nothing there—just wanted to see the GETs produced by LS)

Note the `curl` line in your code wasn’t closed (maybe a copypaste error?).
If you’re using that inside something bigger (like AzuraCast) use `log` instead of `print`.

Also, be aware that with live streams, you often get
- no metadata
- metadata only once in a while (when a new DJ/show starts)
- artist _and_ title metadata in `title` (usually formatted as "artist - title"), while `artist` stays blank

If you run into that situation, maybe part of my code can help—adapt as needed:
```ruby
nowplaying = ref("")

def update_nowplaying(m)
  # For diagnostics: Show all metadata
  #log("META: #{metadata.cover.remove(m)}")
  # If we get an empty artist, assume a stream
  # and try to split the title into artist and title.
  nowplaying.set(shorten(m["title"]) ^ "\n" ^ shorten(m["artist"]))
  if string.length(m["artist"]) <= 0 then
    if string.contains(substring=" - ", m["title"]) then
      let (a, t) = string.split.first(separator=" - ", m["title"])
      nowplaying.set(shorten(t) ^ "\n" ^ shorten(a))
    end
  end
  ...
end
```
You can neglect `shorten()`—it’s just used to shorten long names for display in my video stream.

**Comment by gavin2812** [comment]
Thanks all, I ended up going what would seem a very long way round and the only one I could get to work was the base64.encode function, so this also achieved what I was looking for, so used this instead...

Thanks for the replies.
